### PR TITLE
Cancel writing program progress if the same program is bought from darkweb

### LINKED ISSUE #3234: [Visual] If the program is bought meanwhile it's manually writing, the percentage is still shown
Steps to reproduce:
1) Start writing the program
2) Buy the same program on darkweb
3) Cancel writing the program

Expected behavior:
The manual task should be autocancelled when the program is bought, the progress should be nulled. 

image

### PR DESCRIPTION
**Pull request by SagePtr**
If the player is writing a program and the same program is bought from the darkweb meanwhile, the creation progress cancels.

Fixes #3234 (it doesn't fix already bought programs display, only applies to new)

Also, provides proper fix for #2024 (previous fix only applied if the program was bought via terminal but not via singularity API)

### CODE REVIEW

File: src/NetscriptFunctions/Singularity.ts [hunk 1]
```diff
@@ -555,6 +555,22 @@ export function NetscriptSingularity(
...
       }
 
+      const programsRef = player.getHomeComputer().programs;
+      // Remove partially created program if there is one
+      const existingPartialExeIndex = programsRef.findIndex(
+        (program) => item?.program && program.startsWith(item?.program),
+      );
+      // findIndex returns -1 if there is no match, we only want to splice on a match
+      if (existingPartialExeIndex > -1) {
+        programsRef.splice(existingPartialExeIndex, 1);
```
Comment (phyzical) [inline src/NetscriptFunctions/Singularity.ts:565]: hmm would removing a partial program cause issues if they are working on it? maybe put the cancel work above juust incase

**Comment by phyzical** [comment]
just a gif of it in action :)

**Comment by SagePtr** [comment]
> just a gif of it in action :)

What FOSS software can i use to make a gif?

**Comment by phyzical** [comment]
i use [giphy](https://giphy.com/apps/giphycapture) on mac and i use https://www.screentogif.com/ on windows

**Review by phyzical — APPROVED** [review]
thx for the gif, lgtm

**Comment by hydroflame** [comment]
Closing because I wanted to do some refactor related to this but this change is included.

**Pull request closed without merging**